### PR TITLE
Revit2023

### PR DESCRIPTION
**Pull request by teocomi**
## Description

- adds support for Revit 2023
- introduces quite a few changes in some structural conversion @Reynold-Chan please have a looksie!

## Docs
- Will be updated ASAP

### CODE REVIEW

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs [hunk 1]
```diff
@@ -260,6 +261,172 @@ private Element2D AnalyticalSurfaceToSpeckle(AnalyticalModelSurface revitSurface
...
       return speckleElement2D;
     }
+#else
+    private Element2D AnalyticalSurfaceToSpeckle(AnalyticalPanel revitSurface)
+    {
+
+
+      var speckleElement2D = new Element2D();
+
+      var structuralElement = revitSurface.Document.GetElement(AnalyticalToPhysicalAssociationManager.GetAnalyticalToPhysicalAssociationManager(Doc).GetAssociatedElementId(revitSurface.Id));
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs:271]: I am not sure if this method returns the actual analytical element, to be checked!

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs [hunk 2]
```diff
@@ -260,6 +261,172 @@ private Element2D AnalyticalSurfaceToSpeckle(AnalyticalModelSurface revitSurface
...
+
+        displayLine.segments.Add(CurveToSpeckle(loop));
+
+      }
+
+      speckleElement2D.topology = edgeNodes;
+      speckleElement2D["displayValue"] = displayLine;
+
+
+      //speckleElement2D.voids = voidNodes;
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs:303]: I could not find a way to get the voids or inner loops

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertLocation.cs [hunk 3]
```diff
@@ -71,6 +72,19 @@ private Base TryGetLocationAsCurve(FamilyInstance familyInstance)
           return CurveToSpeckle(analyticalModel.GetCurve()) as Base;
         }
       }
+#else
+      var manager = AnalyticalToPhysicalAssociationManager.GetAnalyticalToPhysicalAssociationManager(Doc);
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertLocation.cs:76]: To be chacked as above if this returns the analytical model

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalStick.cs [hunk 4]
```diff
@@ -48,30 +45,38 @@ public List<ApplicationPlaceholderObject> AnalyticalStickToNative(Element1D spec
...
+    {
+#if !REVIT2023
+      var analyticalModel = (AnalyticalModelStick)element.GetAnalyticalModel();
+      analyticalModel.SetReleases(true, Convert.ToBoolean(element1d.end1Releases.stiffnessX), Convert.ToBoolean(element1d.end1Releases.stiffnessY), Convert.ToBoolean(element1d.end1Releases.stiffnessZ), Convert.ToBoolean(element1d.end1Releases.stiffnessXX), Convert.ToBoolean(element1d.end1Releases.stiffnessYY), Convert.ToBoolean(element1d.end1Releases.stiffnessZZ));
+      analyticalModel.SetReleases(false, Convert.ToBoolean(element1d.end2Releases.stiffnessX), Convert.ToBoolean(element1d.end2Releases.stiffnessY), Convert.ToBoolean(element1d.end2Releases.stiffnessZ), Convert.ToBoolean(element1d.end2Releases.stiffnessXX), Convert.ToBoolean(element1d.end2Releases.stiffnessYY), Convert.ToBoolean(element1d.end2Releases.stiffnessZZ));
+      analyticalModel.SetOffset(AnalyticalElementSelector.StartOrBase, offset1);
+      analyticalModel.SetOffset(AnalyticalElementSelector.EndOrTop, offset2);
+#else
+      var analyticalModel = Doc.GetElement(AnalyticalToPhysicalAssociationManager.GetAnalyticalToPhysicalAssociationManager(Doc).GetAssociatedElementId(element.Id)) as AnalyticalMember;
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalStick.cs:73]: Could not find a way to get and set the offsets in the new API!

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalStick.cs [hunk 5]
```diff
@@ -390,6 +395,313 @@ private Element1D AnalyticalStickToSpeckle(AnalyticalModelStick revitStick)
...
+      }
+
+      prop.profile = speckleSection;
+      prop.material = speckleMaterial;
+      prop.name = stickFamily.Name;
+
+
+      var mark = GetParamValue<string>(stickFamily, BuiltInParameter.ALL_MODEL_MARK);
+
+      //TODO: how to differenciate between column and beam?
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalStick.cs:680]: See comment, seems like there is no difference now?

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs [hunk 6]
```diff
@@ -260,6 +261,172 @@ private Element2D AnalyticalSurfaceToSpeckle(AnalyticalModelSurface revitSurface
 
       return speckleElement2D;
     }
+#else
+    private Element2D AnalyticalSurfaceToSpeckle(AnalyticalPanel revitSurface)
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs:265]: Maybe we can avoid duplicating the entire method?

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalStick.cs (hunk 4 above)
Comment (Reynold-Chan) [reply Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalStick.cs:73]: Seems like they just moved the method out of the analytical model. This offset is only used for context of BIM in particular from a Revit POV. We can use the offset from the base element instead if it's missing

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs (hunk 2 above)
Comment (Reynold-Chan) [reply Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertAnalyticalSurface.cs:303]: Seems like they've separated the modelling completely. There's a new element called openings. These are voids in Revit2023. Probably need to use these methods instead. Not sure if they have a parent element though to attach to for these openings.

**Comment by teocomi** [comment]
Thanks for the comments @Reynold-Chan , let me know if this is good for an early 2023 release and we'll include in the upcoming one. It doesn't have to be perfect, but should support basic structural elements at least...

**Comment by Reynold-Chan** [comment]
It's not supporting basic structural elements right now . So if we want to support revit to ETABS, that's not possible. Just missing the sections.

**Comment by Reynold-Chan** [comment]
fixed and updated

**Comment by teocomi** [comment]
🙌 nice, are you also resolving the conflicts and merging @Reynold-Chan ?

**Comment by Reynold-Chan** [comment]
Yep can do that

**Pull request merged**